opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -297,10 +297,31 @@ namespace Opm
         debug_cost_counter_ = 0;
         // does the well have a THP related constraint?
         const auto& summaryState = ebosSimulator.vanguard().summaryState();
-        if (!Base::wellHasTHPConstraints(summaryState) || bhp_controlled_well) {
+        if (!Base::wellHasTHPConstraints(summaryState)) {
             computeWellRatesAtBhpLimit(ebosSimulator, well_potentials, deferred_logger);
         } else {
-            well_potentials = computeWellPotentialWithTHP(ebosSimulator, deferred_logger);
+            auto bhp_well_potentials = well_potentials;
+            auto thp_well_potentials = well_potentials;
+            computeWellRatesAtBhpLimit(ebosSimulator, bhp_well_potentials, deferred_logger);
+            // TODO: we probably should check whether converged with BHP wether honor THP limit
+            thp_well_potentials = computeWellPotentialWithTHP(ebosSimulator, deferred_logger);
+            const double sum_bhp_well_potentials = std::abs(std::accumulate(bhp_well_potentials.begin(), bhp_well_potentials.end(), 0.0));
+            const double sum_thp_well_potentials = std::abs(std::accumulate(thp_well_potentials.begin(), thp_well_potentials.end(), 0.0));
```

Comment:
Are you sure that this is what you meant or would you rather accumulate the absolute values? sum might be nearly zero despite large absolute potentials.

Reply:
> Are you sure that this is what you meant or would you rather accumulate the absolute values?

I think they should be very similar, either will work. Mathematically, your suggestion is valid. 

> sum might be nearly zero despite large absolute potentials.

If the potentials are off different signs, we have a real problem to tackle.  If a well is producing oil while injecting gas/water, I think a lot of assumptions of our solution strategy will have problems. (It might happen with near-dying wells that have some floating-point calculation errors.)